Documentation for pure Python device interface

mondus opened the pull request:
Adds device side examples for all current CUDA C++. Discussion required about naming of this. I personally don't like CUDA C++ as it implies a user might need to know CUDA. The abstraction is to specifically avoid this so it might be better to use "Agent Function/Device C++" and provide some context as to the fact that this is a subset of C++ supported by CUDA in the docs. Pure Python has also been used to describe the Python device side code. This could potentially also be changed.

Main repo PR: https://github.com/FLAMEGPU/FLAMEGPU2/pull/882

Robadob commented:
Tutorial should be updated too

On Tue, 16 Aug 2022, 17:48 mondus, ***@***.***> wrote:

> Adds device side examples for all current CUDA C++. Discussion required
> about naming of this. I personally don't like CUDA C++ as it implies a user
> might need to know CUDA. The abstraction is to specifically avoid this so
> it might be better to use "Agent Function/Device C++" and provide some
> context as to the fact that this is a subset of C++ supported by CUDA in
> the docs. Pure Python has also been used to describe the Python device side
> code. This could potentially also be changed.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104
> Commit Summary
>
>    - 766f650
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/commits/766f65099da3763659a232f62471ed933a945545>
>    documentation for pure Python device interface
>
> File Changes
>
> (9 files <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files>)
>
>    - *M* src/guide/agent-functions/agent-birth-death.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-36f03979c193ca736f14891318fbfe70f8cd6268c4b37f878c1c6c6eed5a8a7f>
>    (36)
>    - *M* src/guide/agent-functions/agent-communication.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-89005c2a60e4850ecb33529c5b692103d8a2dad0d7e23c73276b55d2e9619123>
>    (225)
>    - *M* src/guide/agent-functions/conditional-behaviours.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-267c547c986f6bb1941ed469b0d553d601adc87759df0d3c0a2f1ba39f4cb0b8>
>    (2)
>    - *M* src/guide/agent-functions/defining-agent-functions.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-95937f05ec7e0d3bbb6265792db152c4b06e10d5af916e3a2d39421b8814177b>
>    (155)
>    - *M* src/guide/agent-functions/index.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-2f8c84a59e90ca661d657842542c5f78a5b8585a4967d0992db07cc657365f17>
>    (2)
>    - *M* src/guide/agent-functions/interacting-with-environment.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-3b93ff770b70b3d1e0f369573196eda625b98e47224d28cf670674beb12c2ef3>
>    (54)
>    - *M* src/guide/agent-functions/modifying-agent-variables.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-bc7817986036b49e92eb40389571eb976f614a97a12ef764ac5a2e648eb038d0>
>    (48)
>    - *M* src/guide/agent-functions/random-numbers.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-b13fc2623e6c783c056a651ef736ab29c3cfa2e4c8e4b814a7d5245213a1c933>
>    (13)
>    - *M* src/guide/defining-messages-communication/index.rst
>    <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104/files#diff-25ace9d84b968795475c02a797cf43bf9435df4dd70a3c04989d8f96c2322434>
>    (1)
>
> Patch Links:
>
>    - https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104.patch
>    - https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/104>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFVGCTAENCVLOA3HQCHJ73VZPA53ANCNFSM56WQ2OCQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+FLAME GPU Pure Python Agent Functions
+-------------------------------------
+
+Ane experimental feature allows the specification of *pure* Python agent functions. In this case agent functions can be expressed using a limited subset of Python. This is limited to syntax which can be transpiled to a C++ function at runtime for compilation. For example Python functionality like generator expressions, arrays, dictionaries, etc, which do not have an obvious C++ equivalent (exclud ...
```
Typo Ane -> An*

Robadob replied:
> to a C++ function
to a C++ equivalent

would make more sense

Robadob replied:
> excluding *the* in the SDL

excluding *the* in the SDL

Not even sure what you mean by SDL here.

mondus replied:
It was supposed to say "including translation that would require use of the C++ STL which has very limited supported in agent functions". Fixed.

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+
+Ane experimental feature allows the specification of *pure* Python agent functions. In this case agent functions can be expressed using a limited subset of Python. This is limited to syntax which can be transpiled to a C++ function at runtime for compilation. For example Python functionality like generator expressions, arrays, dictionaries, etc, which do not have an obvious C++ equivalent (exclud ...
+
+Pure Python agent functions are required to have the ``@pyflamegpu.agent_function`` decorator and must specify two arguments, the message in message out variable, both of which must use a type annotation of a :ref:`supported message type<Communication Strategies>` prefixed with the Python FLAME GPU module name ``pyflamegpu.``.
```
> the message in message out variable

the `message_in` and `message_out` variables

mondus replied:
fixed

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+``math.pi``   Translated of ``M_PI`` in C++
+``math.e``    Translated of ``M_E`` in C++
+``math.inf``  Translated of ``INFINITY`` in C++
+``math.nan``  Translated of ``NAN`` in C++
```
How does Python handle NaN comparison?

NaN can only be tested with `isnan()` in C++.

Robadob replied:
(Or comparing the variable with itself `a!=a` apparently evaluateS true if a is NaN.)

Robadob replied:
```
>>> import math
>>> a = math.nan
>>> a != a
True
```
Applies to Python too, so nan without `isnan` seems pretty useless.

mondus replied:
You could use `isnan` in pythin by calling `isnan(x)` which would translate directly to the same call in C++ (but generate a warning) or `math.isnan` which would translate to `isnan(x)` (with no warning). 

Its not *pure* Python but you its probably worth leaving in as you can generate functional C++ using it.

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+``UInt8``     Would be ``uint_8`` type template argument 
+``Int16``     Would be ``int_16`` type template argument 
+``UInt16``    Would be ``uint_16`` type template argument 
+``Int32``     Would be ``int_32`` type template argument 
```
Might want to point out that `Int32` is the same as `Int` (and same for unsigned)

mondus replied:
Done

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+``UInt64``    Would be ``uint_64`` type template argument 
+============= ===============================================
+
+In each case the Type is appended to the end of the API function. E.g. ``getVariableInt("name")`` in Python would be ``getVariable<"int">("name")`` in C++, and ``getVariableFloat("name")`` would be  ``getVariable<"float">("name")``  in C++
```
Why is type capitalised?

mondus replied:
fixed

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+``numpy.longdouble``   A numpy ``longdouble`` type which will transpile to a C++ ``long double``
+``numpy.bool_``        A numpy ``bool_`` type which will transpile to a C++ ``bool``
+``numpy.bool8``        A numpy ``bool8`` type which will transpile to a C++ ``bool``
+``numpy.int_``         A numpy ``int_`` type which will transpile to a C++ ``long``
```
noooo `long` is bad and should be avoided. It's platform specific. 32 bit windows, 64 bit linux.

mondus replied:
I think that this is designed specifically this way to allow use of platform defined size types. Either way I think we should stick to the types suggested in the numpy docs: https://numpy.org/doc/stable/user/basics.types.html

If someone wants something exactly 32bits then there is always an alternative they can pick.

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+``numpy.half``         A numpy ``half`` type which will transpile to a C++ ``half``
+``numpy.single``       A numpy ``single`` type which will transpile to a C++ ``float``
+``numpy.double``       A numpy ``double`` type which will transpile to a C++ ``double``
+``numpy.longdouble``   A numpy ``longdouble`` type which will transpile to a C++ ``long double``
```
CUDA doesn't support long double afaik.

mondus replied:
Not yet. I'm included to leave it in as this table covers all the numpy data types. I will add a note to say that it is not currently supported in device code.

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+``numpy.int32``        A numpy ``int32`` type which will transpile to a C++ ``int32_t``
+``numpy.int64``        A numpy ``int64`` type which will transpile to a C++ ``int64_t``
+``numpy.intp``         A numpy ``intp`` type which will transpile to a C++ ``intptr_t``
+``numpy.uint_``        A numpy ``uint_`` type which will transpile to a C++ ``long``
```
same long issue.

mondus replied:
Same answer as previous

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,12 +84,128 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+
+Any runtime compiled agent functions must include the definition of any device functions in the agent function string. These can not be shared between agent functions (as each has a unique string definition).
+
+Pure python device functions must be contained in the source module of any agent functions which call them. Device functions are automatically discovered and included in the transpilation process for agent agent functions in the same Python module. 
```
Perhaps clarify that module means file in this context? Or does it just mean has to be imported by the file?

mondus replied:
Have clarified this

Robadob commented on src/guide/agent-functions/interacting-with-environment.rst:
```diff
@@ -48,10 +58,25 @@ Environmental macro properties can be read, via the returned :class:`DeviceMacro
 They can also be updated with a selection of functions, which execute atomically. These functions will update a single variable and return information related to it's old or new state. This can be useful, for simple actions such as conflict resolution and counting. However, if a basic read is subsequently required, a separate host or agent function in a following layer must be used (otherwise the ...
 
-Macro properties support the normal :func:`+<flamegpu::DeviceMacroProperty::operator+>`, :func:`-<flamegpu::DeviceMacroProperty::operator->`, :func:`+=<flamegpu::DeviceMacroProperty::operator+=>`, :func:`-=<flamegpu::DeviceMacroProperty::operator-=>`, :func:`++<flamegpu::DeviceMacroProperty::operator++>`, :func:`--<flamegpu::DeviceMacroProperty::operator-->` operations. They also have access to a ...
+Macro properties support the normal :func:`+<flamegpu::DeviceMacroProperty::operator+>`, :func:`-<flamegpu::DeviceMacroProperty::operator->`, :func:`+=<flamegpu::DeviceMacroProperty::operator+=>`, :func:`-=<flamegpu::DeviceMacroProperty::operator-=>`, :func:`++<flamegpu::DeviceMacroProperty::operator++>` (C++ only supports pre and post increment), :func:`--<flamegpu::DeviceMacroProperty::operator ...
```
>  (C++ only supports pre and post increment)

Is that what you mean to write, shouldn't it be  "only C++..."

Applies twice in the line

mondus replied:
Correct. Have fixed this.

Robadob commented on src/guide/agent-functions/agent-communication.rst:
```diff
@@ -270,10 +401,10 @@ Similar to spatial messaging, array messages provide several iterators for acces
 ================================= =============================================== ==================================
 Iterator                          Usage                                           API Docs
 ================================= =============================================== ==================================
-Moore Neighbourhood               ``FLAMEGPU->message_in(<arguments>)``           :func:`1D<flamegpu::MessageArray::In::operator()>`, :func:`2D<flamegpu::MessageArray2D::In::operator()>`, :func:`3D<flamegpu::MessageArray3D::In::operator()>`
```
I understand why you've done this change, but I don't really like it (or have a better suggestion).

mondus replied:
Have added a separate C++ and Python usage column. Although the API Docs column is unnecessarily large (I think because it is possibly based on the pre-processed URL text rather than the display text). Not a massive issue but a bit annoying.

Robadob reviewed (changes requested):
Tutorial also needs changes, but otherwise I think you've hit everything.

Added several notes to your new content where you've got small typographic issues.

Thanks for the corrections to other stuff throughout.

Re naming, my best suggestion would be `Agent C++`, `Agent Python`.
It's `CUDA C++` because it's technically the CUDA subset of supported C++, but as you say not clear to most users.

Robadob commented:
Noticed some small typographic changes in your updates, have added comments.
Not yet looked through the tutorial changes, will probably check the built version of that too but will take a bit longer to go through.

mondus commented:
Have addressed the additional comments and used Agent C++ and Agent Python throughout.

In the tutorial I have removed the proceeding lines from the code examples and replaced with `... // following on from some.code()`

I have also removed the definition of `main` (but left a comment explaining that all tutorial code should be within a main function). It meant that all example had strange indentation and that there were several places that the closing bracket could be defined depending on optional usage of vis/logging.

Robadob commented on src/tutorial/index.rst:
```diff
@@ -209,7 +209,7 @@ In order to create a :class:`MessageSpatial2D::Description<flamegpu::MessageSpat
 .. note::
   
-    The Python interface does not support C++'s templates and nested classes so there are differences in naming style. In almost all cases the template argument is simply appended to the name.
+    The Python interface does not support C++'s templates and nested classes so there are differences in naming style. :ref:`In almost all cases the template argument is simply appended to the name.<Device Agent Communication>`.
```
I don't understand this link, it doesn't seem relevant.

mondus replied:
Used the wrong link. Corrected it to `Pure Python Types` which has a table of suffix for functions.

Robadob commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+In FLAME GPU 2, agent functions can be implemented using the C++ :c:macro:`FLAMEGPU_AGENT_FUNCTION(name, input_message, output_message)<FLAMEGPU_AGENT_FUNCTION>` macro function. It is expanded by the compiler, to produce the full definition of an agent function (see it's API documentation for an example of it's expansion). However, for our usage we simply need to provide it three parameters; the  ...
 
-We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` is used for the output message argument.
+The C++ format of agent function description can also be used to compile agent functions specified as C++ strings at runtime. This enables models specified in Python to compile on the fly. Runtime compilation adds a small additional cost to the initial execution of an agent function, due to compilation. However, FLAME GPU caches compiled agent functions to remove this for repeated runs (if the ag ...
```
> The C++ format of agent function description can also be used to compile agent functions specified as C++ strings at runtime. This enables models specified in Python to compile on the fly. 

This reads badly, reword

mondus replied:
Fixed

Robadob commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+
+To describe our behaviour, we will start by implementing the agent function, whereby each agent outputs a message sharing their location.
+
+We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` (``pyflamegpu.MessageNone in Python``) is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` (``pyflamegpu.MessageSpatial2D in Python``) is used for the outpu ...
```
in Python should be part of the inline code block x2

Also do you not mean pure-python here.

mondus replied:
I have called this Agent Python and made a few changes throughout the document as well as clarifying that *pure* python is a subset of Python with this name.

Robadob replied:
See new comment, you left one in code tags.

Robadob commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+To read an agent's variables the :func:`FLAMEGPU->getVariable()<template<typename T, unsigned int N> __device__ T flamegpu::DeviceAPI::getVariable(const char(&)[N]) const>` function is used in C++. As you may expect by now, the variable's type must be passed as a template argument, and it's name is the only argument. To read an agent's ID, :func:`FLAMEGPU->getID()<flamegpu::DeviceAPI::getID>` is  ...
 
-Functionality for the message output is accessed via ``FLAMEGPU->message_out``, this object is specialised depending on the output message type originally specified in the :c:macro:`FLAMEGPU_AGENT_FUNCTION<FLAMEGPU_AGENT_FUNCTION>` macro. The spatial 2D specialisation, :class:`flamegpu::MessageSpatial2D::Out`, has two available functions; :func:`setVariable()<template<typename T, unsigned int N>  ...
+Functionality for the message output is accessed via ``FLAMEGPU->message_out`` (or named ``message_out`` variable in Python), this object is specialised depending on the output message type originally specified in the :c:macro:`FLAMEGPU_AGENT_FUNCTION<FLAMEGPU_AGENT_FUNCTION>` macro (or via the Python type annotation). The spatial 2D specialisation, :class:`flamegpu::MessageSpatial2D::Out`, has t ...
```
should be `int` type rather than integer type, for consistency.

mondus replied:
fixed

Robadob replied:
Done rst code tags wrong, see new comment

Robadob commented on src/tutorial/index.rst:
```diff
@@ -541,10 +533,52 @@ The remainder of the Circles model's message input agent function contains some
+        x1 = pyflamegpu.getVariableFloat("x")
+        y1 = pyflamegpu.getVariableFloat("y")
+        count = 0
+        for (message : message_in(x1, y1)) :
```
that's not a python style loop?

mondus replied:
There were a few code typos I found when putting all this in to the full example. I thought I had ported them all back into the various places but must have missed this one. Now fixed.

Robadob commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+Functionality for the message output is accessed via ``FLAMEGPU->message_out`` (or named ``message_out`` variable in Python), this object is specialised depending on the output message type originally specified in the :c:macro:`FLAMEGPU_AGENT_FUNCTION<FLAMEGPU_AGENT_FUNCTION>` macro (or via the Python type annotation). The spatial 2D specialisation, :class:`flamegpu::MessageSpatial2D::Out`, has t ...
 
-Finally, all agent functions must return either :enumerator:`flamegpu::ALIVE<flamegpu::AGENT_STATUS::ALIVE>` or :enumerator:`flamegpu::DEAD<flamegpu::AGENT_STATUS::DEAD>`. Unless the agent function is specified to support agent death inside the :class:`AgentFunctionDescription<flamegpu::AgentFunctionDescription>` via :func:`setAllowAgentDeath()<flamegpu::AgentFunctionDescription::setAllowAgentDea ...
+Finally, all agent functions must return either :enumerator:`flamegpu::ALIVE<flamegpu::AGENT_STATUS::ALIVE>` or :enumerator:`flamegpu::DEAD<flamegpu::AGENT_STATUS::DEAD>` (``pyflamegpu.ALIVE`` or ``pyflamegpu.DEAD`` respectively in Python). Unless the agent function is specified to support agent death inside the :class:`AgentFunctionDescription<flamegpu::AgentFunctionDescription>` via :func:`setA ...
```
"in Python)" should this not be "pure python"

Robadob replied:
or agent python

mondus replied:
changed to agent python

Robadob commented on src/tutorial/index.rst:
```diff
@@ -435,30 +421,40 @@ Below you can see how the message output function may be assembled. Normally, ag
+    @pyflamegpu.agent_function
+    def output_message(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageSpatial2D):
+        message_out.setVariableUInt("id", pyflamegpu.getID())
+        message_out.setLocation( \
```
Are these slashes actually required?

mondus replied:
Actually no and Pep suggests not to so removed.

Robadob commented on src/tutorial/index.rst:
```diff
@@ -435,30 +421,40 @@ Below you can see how the message output function may be assembled. Normally, ag
 Each FLAME GPU message type provides unique methods for accessing messages, in this case we are using the :class:`MessageSpatial2D<flamegpu::MessageSpatial2D>` type. Refer to the :ref:`agent communication guide<Device Agent Communication>` for details of other messaging format's usage.
 
-The only way to access spatial messaging types is via an iterator, which returns all messages in a Moore neighbourhood about the provided search location. This means, that all messages within the originally specified search radius will be returned, however it is necessary for the user to filter out the small number of messages which fall outside of this radius that will also be returned. Furtherm ...
+The only way to access spatial messaging types is via an iterator, which returns all messages in a Moore neighbourhood (discretised by the message radius) about the provided search location. This means, that all messages within the originally specified search radius will be returned, however it is necessary for the user to filter out messages which are contained within the Moor neighbour but fall ...
```
Typo "Moor"

Robadob replied:
May also reword slightly the "that are also returned" at the end of the sentence doesn't read nicely.

Robadob replied:
Original typo "agent's" should be "agents"

Robadob replied:
outputting ... to the message reads weird, maybe swap for attaching, also change "agent id" to "agent's id"

mondus replied:
Have tidied all of this points up.

mondus commented:
Also rebased from master. Hope it hasn't knackard the comment lines.

Robadob commented on src/tutorial/index.rst:
```diff
@@ -435,30 +421,40 @@ Below you can see how the message output function may be assembled. Normally, ag
+The only way to access spatial messaging types is via an iterator, which returns all messages in a Moore neighbourhood (discretised by the message radius) about the provided search location. This means, that all messages within the originally specified search radius will be returned, however it is necessary for the user to filter out messages which are contained within the Moor neighbour but fall ...
 
-The spatial message iterator is accessed using :func:`FLAMEGPU->message_in()<flamegpu::MessageSpatial2D::In::operator()>`, this takes two ``float`` parameters specifying the search origin. Normally this will be passed directly to a C++ range-based for loop (e.g. ``or (const auto &message : FLAMEGPU->message_in(my_x, my_y)) {``), allowing the returned messages to be iterated.
+The spatial message iterator is accessed using :func:`FLAMEGPU->message_in()<flamegpu::MessageSpatial2D::In::operator()>` (or via the ``message_in`` agent function argument in Python), this takes two ``float`` parameters specifying the search origin. Normally this will be passed directly to a C++ range-based for loop, allowing the returned messages to be iterated.
```
Pure python/agent python?

mondus replied:
agent python

mondus replied:
For consistency I have used *pure python* to describe the fact that the functions are written is a subset of *pure* python (as opposed to Python with C++ strings) but consistently named the subset of Python as *Agent Python*

Robadob commented on src/tutorial/index.rst:
```diff
@@ -435,30 +421,40 @@ Below you can see how the message output function may be assembled. Normally, ag
+The spatial message iterator is accessed using :func:`FLAMEGPU->message_in()<flamegpu::MessageSpatial2D::In::operator()>` (or via the ``message_in`` agent function argument in Python), this takes two ``float`` parameters specifying the search origin. Normally this will be passed directly to a C++ range-based for loop, allowing the returned messages to be iterated.
 
-In the case of :class:`MessageSpatial2D<flamegpu::MessageSpatial2D>`, the returned :class:`Message<flamegpu::MessageSpatial2D::In::Filter::Message>` objects only provide :func:`getVariable()<template<typename T, unsigned int N> __device__ T flamegpu::MessageSpatial2D::In::Filter::Message::getVariable(const char(&)[N]) const>` methods for returning the variables and array variables stored within t ...
+In the case of :class:`MessageSpatial2D<flamegpu::MessageSpatial2D>`, the returned :class:`Message<flamegpu::MessageSpatial2D::In::Filter::Message>` objects only provide :func:`getVariable()<template<typename T, unsigned int N> __device__ T flamegpu::MessageSpatial2D::In::Filter::Message::getVariable(const char(&)[N]) const>` methods for returning the variables and array variables stored within t ...
```
>  The *pure* Python equivalent requires appended Types and array lengths

requires the type and array length to be appended to the name?

mondus replied:
fixed

Robadob commented on src/tutorial/index.rst:
```diff
@@ -435,30 +421,40 @@ Below you can see how the message output function may be assembled. Normally, ag
+In the case of :class:`MessageSpatial2D<flamegpu::MessageSpatial2D>`, the returned :class:`Message<flamegpu::MessageSpatial2D::In::Filter::Message>` objects only provide :func:`getVariable()<template<typename T, unsigned int N> __device__ T flamegpu::MessageSpatial2D::In::Filter::Message::getVariable(const char(&)[N]) const>` methods for returning the variables and array variables stored within t ...
 
-Accessing environment properties is very similar to accessing agent and message variables, :func:`getProperty()<template<typename T, unsigned int N> T flamegpu::ReadOnlyDeviceEnvironment::getProperty(const char(&)[N]) const>` is called on :class:`FLAMEGPU->environment<flamegpu::DeviceEnvironment>`.
+Accessing environment properties is very similar to accessing agent and message variables, :func:`getProperty()<template<typename T, unsigned int N> T flamegpu::ReadOnlyDeviceEnvironment::getProperty(const char(&)[N]) const>` is called on :class:`FLAMEGPU->environment<flamegpu::DeviceEnvironment>`. The *pure* Python equivalent requires appended Types and array lengths.
```
See above

mondus replied:
fixed

Robadob commented on src/guide/agent-functions/agent-communication.rst:
```diff
@@ -290,9 +421,9 @@ Below are some examples using each of the iterators:
+    .. code-tab:: cuda Agent C++ Moore
 
-      // Define an agent function, "inputdata" which has accepts an input message using the "MessageSpatial3D" communication strategy and inputs no messages
+      // Define an agent function, "inputdata" which has accepts an input message using the "MessageArray3D" communication strategy and outputs no messages
```
Original typo
> has accepts

mondus replied:
fixed in all three examples

Robadob commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,38 +84,165 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+
+Any runtime compiled agent functions must include the definition of any device functions in the agent function string. These can not be shared between agent functions (as each has a unique string definition).
+
+Pure python device functions must be contained in the source file of any agent functions which call them. All device functions are automatically discovered and included in the transpilation process. 
```
>in the transpilation

during the...*

mondus replied:
agree

Robadob commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+
+To describe our behaviour, we will start by implementing the agent function, whereby each agent outputs a message sharing their location.
+
+We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` (``pyflamegpu.MessageNone in Agent Python``) is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` (``pyflamegpu.MessageSpatial2D`` in Agent Python) is used f ...
```
>  in Agent Python

Still inside code tags in one instance

mondus replied:
got it

Robadob commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+To read an agent's variables the :func:`FLAMEGPU->getVariable()<template<typename T, unsigned int N> __device__ T flamegpu::DeviceAPI::getVariable(const char(&)[N]) const>` function is used in C++. As you may expect by now, the variable's type must be passed as a template argument, and it's name is the only argument. To read an agent's ID, :func:`FLAMEGPU->getID()<flamegpu::DeviceAPI::getID>` is  ...
 
-Functionality for the message output is accessed via ``FLAMEGPU->message_out``, this object is specialised depending on the output message type originally specified in the :c:macro:`FLAMEGPU_AGENT_FUNCTION<FLAMEGPU_AGENT_FUNCTION>` macro. The spatial 2D specialisation, :class:`flamegpu::MessageSpatial2D::Out`, has two available functions; :func:`setVariable()<template<typename T, unsigned int N>  ...
+Functionality for the message output is accessed via ``FLAMEGPU->message_out`` (or named ``message_out`` variable in Agent Python), this object is specialised depending on the output message type originally specified in the :c:macro:`FLAMEGPU_AGENT_FUNCTION<FLAMEGPU_AGENT_FUNCTION>` macro (or via the Python type annotation). The spatial 2D specialisation, :class:`flamegpu::MessageSpatial2D::Out`, ...
```
Code tags require double `` in rst (int)

mondus replied:
Ha. Thats from switching between rst and github md.

Robadob reviewed (approved):
Just make sure you squash it

ptheywood commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -3,14 +3,19 @@
-An agent function is defined using the :c:macro:`FLAMEGPU_AGENT_FUNCTION` macro. 
+An experimental feature for Python allows the specification of *pure* agent functions in Python (referred to as *Agent Python* in this guide). The Python can then be transpiled, a process of translating the Python syntax to equivalent C++, at runtime.
+
+C++ Compile Time Agent Functions
```
RTC is technically "compile time" too (Run time **Compiled**). 

The NVRTC docs refer to traditional non-RTC CUDA compilation as `offline static compilation`. This is being very very pedantic, so may not warrent changing but we should be careful with ambiguous language.

mondus replied:
Not sure I want to try and change this at this point but I get your point

ptheywood commented on src/guide/agent-functions/index.rst:
```diff
@@ -8,7 +8,7 @@ As FLAME GPU is an agent-based modelling framework, agent functions are central
 
 Each agent function in FLAME GPU 2 is associated with a particular agent type and is represented by an :class:`AgentFunctionDescription<flamegpu::AgentFunctionDescription>` object. This object describes the name of the function, the agent state in which it should be active, and any state transitions it should apply to the agent. The implementation code for the behaviour is written separately and  ...
-:c:macro:`FLAMEGPU_AGENT_FUNCTION` macro to label the code. Alternatively, the implementations can be written in strings and loaded at runtime. The latter method is the only method supported by FLAME GPU 2's Python API.
+:c:macro:`FLAMEGPU_AGENT_FUNCTION` macro to label the code. Alternatively, the implementations can be written in strings and loaded at runtime. The latter method is supported by FLAME GPU 2's Python API which also includes a feature to permit pure Python agent functions to be specified which are transpiled to equivalent C++ at runtime.
```
The new sentence doesn't add anything to the point that the `FALMEGPU_AGENT_FUNCTION` macro is used. 
It would be better to rewrite this from `The implementation code for the behaviour` to the end, stating that:

The implementation is specified separately, in one of 3 ways:
* As a globally scoped (CUDA) C++ function, labelled with the `FALMEGPU_AGENT_FUNCTION` macro 
* Can be written as a `FALMEGPU_AGENT_FUNCTION` labelled C++ function written as a string, in either a C++ or Python 3 model
* As a (global? file?)-scoped python 3 function annoted as `@pyflamegpu.agent_function` which are transpiled to equivalent C++ and which is compiled at runtime 
I assume there is probably a similar change required for agent function conditions, which just hasn't shown up in the review due to the current lack of support (which should be fixed).

mondus replied:
changed to a list format

ptheywood commented on src/guide/agent-functions/interacting-with-environment.rst:
```diff
@@ -48,10 +58,25 @@ Environmental macro properties can be read, via the returned :class:`DeviceMacro
 They can also be updated with a selection of functions, which execute atomically. These functions will update a single variable and return information related to it's old or new state. This can be useful, for simple actions such as conflict resolution and counting. However, if a basic read is subsequently required, a separate host or agent function in a following layer must be used (otherwise the ...
 
-Macro properties support the normal :func:`+<flamegpu::DeviceMacroProperty::operator+>`, :func:`-<flamegpu::DeviceMacroProperty::operator->`, :func:`+=<flamegpu::DeviceMacroProperty::operator+=>`, :func:`-=<flamegpu::DeviceMacroProperty::operator-=>`, :func:`++<flamegpu::DeviceMacroProperty::operator++>`, :func:`--<flamegpu::DeviceMacroProperty::operator-->` operations. They also have access to a ...
+Macro properties support the normal :func:`+<flamegpu::DeviceMacroProperty::operator+>`, :func:`-<flamegpu::DeviceMacroProperty::operator->`, :func:`+=<flamegpu::DeviceMacroProperty::operator+=>`, :func:`-=<flamegpu::DeviceMacroProperty::operator-=>`, :func:`++<flamegpu::DeviceMacroProperty::operator++>` (only C++ supports pre and post increment), :func:`--<flamegpu::DeviceMacroProperty::operator ...
```
```suggestion
Macro properties support the normal :func:`+<flamegpu::DeviceMacroProperty::operator+>`, :func:`-<flamegpu::DeviceMacroProperty::operator->`, :func:`+=<flamegpu::DeviceMacroProperty::operator+=>`, :func:`-=<flamegpu::DeviceMacroProperty::operator-=>`, :func:`++<flamegpu::DeviceMacroProperty::operator++>` (only C++ supports pre and post increment), :func:`--<flamegpu::DeviceMacroProperty::operator-->` (only C++ supports pre and post decrement) operations. They also have access to a limited set of additional functions, explained in the table below.
```

ptheywood replied:
Could make more terse additionaly via `C++ only` or similar, given they are next to the associated operators.

ptheywood commented on src/guide/agent-functions/interacting-with-environment.rst:
```diff
@@ -93,9 +118,32 @@ Example usage is shown below:
     
 .. warning::
-  Be careful when using :class:`DeviceMacroProperty<flamegpu::DeviceMacroProperty>`. When you retrieve an element e.g. ``location[0][0][0]`` (from the example above), it is of type :class:`DeviceMacroProperty<flamegpu::DeviceMacroProperty>` not ``unsigned int``. Therefore you cannot pass it directly to functions which take generic arguments such as ``printf()``, as it will be interpreted incorrec ...
+  Be careful when using :class:`DeviceMacroProperty<flamegpu::DeviceMacroProperty>`. When you retrieve an element e.g. ``location[0][0][0]`` (from the example above), it is of type :class:`DeviceMacroProperty<flamegpu::DeviceMacroProperty>` not ``unsigned int``. Therefore you cannot pass it directly to functions which take generic arguments such as ``printf()``, as it will be interpreted incorrec ...
```
is `int` correct given the example is for an unsigned integer, so `numpy.uint`?

mondus replied:
Ahh yes. `int` is a catchall for all integer types in Python and the `int()` function is built in. There is no `uint()` equivalent.

ptheywood replied:
hence the suggestion to use `numpy.uint` as previously refferred to in other sections of the docs. 

Surely casting to `int` will impliclty convert the value into a signed integer of sufficient width? 

If you don't want to veryify this behaves correctly by specifying an integer `> INT_MAX` but `< UINT_MAX` then changing the example here to instead use an `int` instead of an `unsigned int` would be the easy but less helpful fix.

mondus replied:
ok have moved it to a a `numpy.uint` (I don't think its strictly necessary but doesn't hurt)

ptheywood commented on src/guide/agent-functions/modifying-agent-variables.rst:
```diff
@@ -31,12 +32,33 @@ with the local copy.
+  .. code-tab:: py Agent Python
+
+    @pyflamegpu.agent_function
+    def ExampleFn(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageNone):
```
Why Upper Cammel case function name?

mondus replied:
Because that is how is was before

ptheywood commented on src/tutorial/index.rst:
```diff
@@ -169,27 +169,27 @@ Before the model description, we will also define two (constant) variables, to d
 
-    #include "flamegpu/flamegpu.h"
+    ...
+    // all code examples are assumed to be implemented within a main function. E.g. int main(int argc, const char **argv)
```
Convention is `const char * argv[]`. Would be good to split over two lines to avoid horizontal scroll bars for narrow windows (e.g. phones).

ptheywood replied:
```suggestion
    // All code examples are assumed to be implemented within a main function.
    // E.g. int main(int argc, const char *argv[])
```

ptheywood commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+In FLAME GPU 2, agent functions can be implemented using the C++ :c:macro:`FLAMEGPU_AGENT_FUNCTION(name, input_message, output_message)<FLAMEGPU_AGENT_FUNCTION>` macro function. It is expanded by the compiler, to produce the full definition of an agent function (see it's API documentation for an example of it's expansion). However, for our usage we simply need to provide it three parameters; the  ...
 
-We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` is used for the output message argument.
+The C++ format of agent function description can be compiled at runtime by specifying the function as a C++ string. This enables models specified in Python to compile on the fly. Runtime compilation adds a small additional cost to the initial execution of an agent function, due to compilation. However, FLAME GPU caches compiled agent functions to remove this for repeated runs (if the agent functi ...
```
```suggestion
The C++ format of agent function description can be compiled at runtime by specifying the function as a C++ string. This enables models specified in Python to compile on the fly. Runtime compilation adds a small additional cost to the initial execution of an agent function, due to compilation. However, FLAME GPU caches compiled agent functions to remove this for repeated runs (if the agent function/model has not changed).

When using Python it is possible to specify agent functions using the C++ format as well as via a *pure* Python (a subset of Python referred to as *Agent Python*) description which is shown in this tutorial. Agent function in *pure* Python must be defined as having a ``@pyflamegpu.agent_function`` decorator and using the following syntax ``def outputdata(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageNone):`` which includes the specifiation of the name and type (uysing type annotations) of the output and input message. The *pure* Python implementation will translate the Python to C++ at runtime prior to compilation through a process known as transpiling.
```

Splitting the very larger paragraph improves readability.

ptheywood commented on src/tutorial/index.rst:
```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
+In FLAME GPU 2, agent functions can be implemented using the C++ :c:macro:`FLAMEGPU_AGENT_FUNCTION(name, input_message, output_message)<FLAMEGPU_AGENT_FUNCTION>` macro function. It is expanded by the compiler, to produce the full definition of an agent function (see it's API documentation for an example of it's expansion). However, for our usage we simply need to provide it three parameters; the  ...
 
-We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` is used for the output message argument.
+The C++ format of agent function description can be compiled at runtime by specifying the function as a C++ string. This enables models specified in Python to compile on the fly. Runtime compilation adds a small additional cost to the initial execution of an agent function, due to compilation. However, FLAME GPU caches compiled agent functions to remove this for repeated runs (if the agent functi ...
```
It might be nicer to use a multi-line code block for the annotated function definition rather than two inline definiations. 

Also mentions `pure python` still.

mondus replied:
It breaks up the flow of the description though. The multi line version is in the full agent function example so I'd prefer to leave this as is.

ptheywood commented on src/tutorial/index.rst:
```diff
@@ -435,30 +421,40 @@ Below you can see how the message output function may be assembled. Normally, ag
+The spatial message iterator is accessed using :func:`FLAMEGPU->message_in()<flamegpu::MessageSpatial2D::In::operator()>` (or via the ``message_in`` agent function argument in Agent Python), this takes two ``float`` parameters specifying the search origin. Normally this will be passed directly to a C++ range-based for loop, allowing the returned messages to be iterated.
 
-In the case of :class:`MessageSpatial2D<flamegpu::MessageSpatial2D>`, the returned :class:`Message<flamegpu::MessageSpatial2D::In::Filter::Message>` objects only provide :func:`getVariable()<template<typename T, unsigned int N> __device__ T flamegpu::MessageSpatial2D::In::Filter::Message::getVariable(const char(&)[N]) const>` methods for returning the variables and array variables stored within t ...
+In the case of :class:`MessageSpatial2D<flamegpu::MessageSpatial2D>`, the returned :class:`Message<flamegpu::MessageSpatial2D::In::Filter::Message>` objects only provide :func:`getVariable()<template<typename T, unsigned int N> __device__ T flamegpu::MessageSpatial2D::In::Filter::Message::getVariable(const char(&)[N]) const>` methods for returning the variables and array variables stored within t ...
```
Give an example for the python case of type name and length. Also mentions pure python still

mondus replied:
Example added

ptheywood commented on src/tutorial/index.rst:
```diff
@@ -541,10 +533,52 @@ The remainder of the Circles model's message input agent function contains some
+                    fx += k * x21
+                    fy += k * y21
+                    count += 1
+                
```
Why so many blank lines?

mondus replied:
No idea. Fixed

ptheywood commented on src/tutorial/index.rst:
```diff
@@ -1256,6 +1280,172 @@ If you have followed the complete tutorial, you should now have the following co
+                    fx += k * x21
+                    fy += k * y21
+                    count += 1
+
+
+
```
No need for 3 blank lines here.

mondus commented on src/guide/agent-functions/conditional-behaviours.rst:
```diff
@@ -7,15 +7,15 @@ Agent function conditions are executed by all agents in the input state before t
 to the end state, agents which return ``false`` fail the condition, do not execute the agent function and remain in the input state.
 
-Agent function conditions are specified using the :c:macro:`FLAMEGPU_AGENT_FUNCTION_CONDITION`, this differs from the normal agent function macro as only the function condition name must be specified.
+Agent function conditions are specified using the :c:macro:`FLAMEGPU_AGENT_FUNCTION_CONDITION` in C++, this differs from the normal agent function macro as only the function condition name must be specified. Agent function conditions are not supported in the pure Python format (used for agent functions) but can be specified as C++ strings within the Python interface.
```
This is left as an enhancement in issue FLAMEGPU2#909

mondus commented on src/guide/agent-functions/agent-communication.rst:
```diff
@@ -290,9 +421,9 @@ Below are some examples using each of the iterators:
 
 .. tabs::
```
Uses a wider screen. Problem solved!

mondus replied:
I've separated these.

mondus commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -3,14 +3,19 @@
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
-Agent Functions can be specified as CUDA functions, built at compile time when using the C++ API, or they can be specified as Run-Time Compiled (RTC) function strings when using the both the C++ and Python APIs.
+Agent Functions can be specified as C++ functions, built at compile time when using the C++ API, or they can be specified as Run-Time Compiled (RTC) function strings when using the both the C++ and Python APIs. Although agent functions are technically CUDA device code (which supports only a subset of C++) the FLAME GPU API abstracts this and no CUDA syntax is required to script behaviour. Agent f ...
```
Have clarified this.

mondus commented on src/guide/agent-functions/defining-agent-functions.rst:
```diff
@@ -74,38 +84,165 @@ When using the Run-Time Compiled (RTC) functions, optionally in the C++ API or r
+
+.. _Pure Python Agent Functions:
+
+FLAME GPU Pure Python Agent Functions
```
Have resolved this according to suggestion on slack

Robadob reviewed (approved):
Already approved this before, when you addressed my points.

GH is showing a merge conflict (over empty lines), so you may want to do a rebase/squash.

ptheywood commented on src/guide/agent-functions/index.rst:
```diff
@@ -7,8 +7,12 @@ As FLAME GPU is an agent-based modelling framework, agent functions are central
-:c:macro:`FLAMEGPU_AGENT_FUNCTION` macro to label the code. Alternatively, the implementations can be written in strings and loaded at runtime. The latter method is the only method supported by FLAME GPU 2's Python API.
+Each agent function in FLAME GPU 2 is associated with a particular agent type and is represented by an :class:`AgentFunctionDescription<flamegpu::AgentFunctionDescription>` object. This object describes the name of the function, the agent state in which it should be active, and any state transitions it should apply to the agent. 
+
+The implementation code for the behaviour (the actual agent function) can be specified in three alternative ways;
```
List does not render correctly due to missing preceding new line

![image](https://user-images.githubusercontent.com/628937/186685731-22a25f6c-64da-44fd-923c-ab682bf1374e.png)
```suggestion
The implementation code for the behaviour (the actual agent function) can be specified in three alternative ways;

```

ptheywood commented on src/guide/agent-functions/index.rst:
```diff
@@ -7,8 +7,12 @@ As FLAME GPU is an agent-based modelling framework, agent functions are central
+The implementation code for the behaviour (the actual agent function) can be specified in three alternative ways;
+* As a globally scoped (CUDA) C++ function, labelled with the :c:macro:`FLAMEGPU_AGENT_FUNCTION` macro
+* As a C++ function labelled with teh :c:macro:`FLAMEGPU_AGENT_FUNCTION` macro but written as a string (supported in either a C++ or Python 3)
+* As a global scoped Python 3 function annotated as `@pyflamegpu.agent_function` which will be transpiled to equivalent C++ and which is compiled at runtime
```
RST requires double backticks for inline code blocks.

```suggestion
* As a global scoped Python 3 function annotated as ``@pyflamegpu.agent_function`` which will be transpiled to equivalent C++ and which is compiled at runtime
```

ptheywood reviewed (changes requested):
Couple of RST issues with suggestions, plus the one outstanding issue about what happens if the (device) python code includes casting an `<unsinged int>` via `int()` which is used in an example.

ptheywood reviewed (approved):
Good for a squash merge after the FLAMEGPU2 PR has been merged.
